### ISSUE TITLE
Correct Projection Issues For CVM-S4.26

### ISSUE THREAD

**Issue by scecdavid** [SCECcode/UCVM#13]
- Reason:
    - Original Fortran conversion code by Chen and Lee had bug in it whereby new easting and northing for the top-left-most corner were calculated incorrectly. The easting was changed first and then the new northing was calculated off of the changed easting not the original easting. Therefore the northing was off.
    - Original lat, lon to UTM code from Chen and Lee included a simplification in the project code borrowed from John Snyder, Map Projections Used By The USGS (https://pubs.usgs.gov/bul/1532/report.pdf), which listed a finite set of terms as a proxy for the infinite series. This led to errors in the UTM coordinates on the order of 70m. This is a systemic offset mostly, although the offset grows as you get further and further away from the equator. So the offset may be 70m at 40 latitude but only 60m at 30 latitude, e.g..
    - Science side: be accurate with projections; software: remove Fortran dependency

- Steps to Fix:
    - Converted Po and En-Jui’s FD_GRID Fortran script into Python
    - Script is located at ucvm/bin/internal/_ucvm_calculate_box_for_inversion
    - New script uses Proj.4 instead of Po and En-Jui’s SPECFEM Fortran code

- Findings:
    - Some slight changes to corners both in UTM and corresponding lat/lorn
    - Off less than 100m in both X and Y
    - Recommended removing Fortran code, using Proj.4

- Decisions Made:
    - Switched buggy Fortran code to Proj.4 DG Yes CG Not Verified
    - Adjusted four corners for CVM-S4.26 using _ucvm_calculate_box_for_inversion

- Action Items:
    - Ensure that all future tomography inversions use Proj.4, not the old Fortran code

**Comment by scecdavid** [comment]
_ucvm_calculate_box_for_inversion script is finished. 

TL;DR see http://hypocenter.usc.edu/research/ucvm/gitissues/github_13.tar.gz for the grid points and output.

I have run it with both the CVM-S4.26 and CCA input files. The grids that I get back appear to be good. The script:

- Calculates the four corners according to PC, EL method but with Proj.4 not Fortran code
- Verifies the distances of the BL, BR and TL, TR side are equal, and ensures that the distances between the BL, TL and BR, TR sides are equal
- Verifies that the rotation angles of the BL, BR point and TL, BL (the sides) match so that way it is a perfect rectangle
- Generates the grid using Proj.4 using a simple point rotation and adding it to the origin
- Verifies that the four corners MATCH with the four corners that were calculated before

I have attached the outputs of running the commands.

Steps to Reproduce:

Execute python3 _ucvm_calculate_box_for_inversion -f CCA06_box.inp to get the CCA box and python3 _ucvm_calculate_box_for_inversion -f CVMS426_box.inp for the CVM-S4.26 box. You will see the calculation output and also the data files included will be generated.

**Comment by scecdavid** [comment]
This grid has been sent off to the scientists for use. I am considering this to be closed.

**Comment by scecdavid** [comment]
Added in capability for the script to merge the downloaded Mira data with the newly generated grid. The resulting file is X-fast. The final columns are:

X, Y, Z, UTM X, UTM Y, LON, LAT, DEPTH, VP, VS, DENSITY

This represents the final "master model" for CVM-S4.26.

**Comment by scecdavid** [comment]
This file has now been uploaded to Mira at /gpfs/mira-home/davidgil/cvms426_final_model.txt.

To regenerate this file, download the original S4.26 tomographic slices from Mira locally (e.g. to ~/Tomography). Then run the command:

_ucvm_calculate_box_for_inversion -f CVMS426_box.inp -d ~/Tomography

This will output a file called generated_grid.txt which matches the final model.

Commit f8aa5a4f91953c67188847e329a7220c0c1cf479 added this capability in.

I am considering this ticket closed again.
